### ISSUE TITLE
from_pretrained() does not update configuration in exp_manager

### ISSUE THREAD

**Issue by piraka9011** [NVIDIA/NeMo#3372]
**Describe the bug**

When fine tuning from a NeMo model (ex. stt_en_cn1024), the exp_manager's cfg is not updated properly.
I can see that in my run the model uses one config, but WandB reports another.

This issue did not occur in v1.4.0 and happened after I upgraded to v1.5.0.
Maybe it has to do with order of operations? See below.

**Steps/Code to reproduce bug**

```py
import pytorch_lightning as pl
from nemo.collections.asr.models import EncDecCTCModelBPE
from nemo.core.config import hydra_runner
from nemo.utils.exp_manager import exp_manager

@hydra_runner(config_path="conf/citrinet/", config_name="config")
def main(cfg):
    trainer = pl.Trainer(**cfg.trainer)
    log_dir = exp_manager(trainer, cfg.get("exp_manager", None))
    asr_model = EncDecCTCModelBPE.from_pretrained(model_name=cfg.init_from_pretrained_model)
    asr_model.encoder.unfreeze()
    asr_model.change_vocabulary(
      new_tokenizer_dir=cfg.model.tokenizer.dir,
      new_tokenizer_type=cfg.model.tokenizer.type
    )
    asr_model.setup_optimization(cfg.model.optim)
    asr_model.setup_training_data(cfg.model.train_ds)
    asr_model.setup_multiple_validation_data(cfg.model.validation_ds)
    asr_model.spec_augmentation = asr_model.from_config_dict(cfg.model.spec_augment)
    asr_model.set_trainer(trainer)
    trainer.fit(asr_model)
```

**Expected behavior**

WandB cfg should display the proper config ([Pastebin of the WandB config](https://pastebin.com/Bc29P9FL))

**Environment overview (please complete the following information)**

 - Environment location: Docker (nvcr.io/nvidia/pytorch:21.10-py3) on AWS EC2 using `docker run -it bash <image>`
 - Method of NeMo install: `pip install nemo_toolkit[asr]==1.5.1`

**Additional context**

GPU model: V100
Nvidia driver: 460

**Comment by titu1994** [comment]
In general, the set_trainer() should be the first call after creating the model - or pass the trainer in the restore_from() or from_pretrained() method itself. 

However this should not affect exp manager - the only thing it depends on is the config passed at this line `log_dir = exp_manager(trainer, cfg.get("exp_manager", None))`. Could you print out this config (cfg.exp_manager) and check whether it matches what you expect ?

**Comment by piraka9011** [comment]
My `exp_manager` cfg:

```yaml
exp_manager:
  exp_dir: "experiments"
  name: *name
  create_tensorboard_logger: true
  create_checkpoint_callback: true
  checkpoint_callback_params:
    monitor: "val_wer"
    mode: "min"
    save_top_k: 2
    always_save_nemo: true
  create_wandb_logger: true
  wandb_logger_kwargs:
    job_type: train
    name: *name
    entity: $WANDB_ENTITY
    project: nemo-experiments
    tags: ["CitriNet-Lg", "train"]
```

**Comment by titu1994** [comment]
Oh I think I got the issue, the config passed to the main method is a new config, but the model is restored by the old config embedded in the Nemo file - you should be able to do model.cfg = cfg before your train step to correctly update the internal config of the model. That's a good catch, I believe only one of the notebooks shows this. Might as well add to the documentation. 

Still, trainer should be set before calling any other function, or passed into the restore_from or from_pretrained call itself

**Comment by piraka9011** [comment]
Sounds good, I will try this out and report back if the proposed solution works.
Will modifying the model's internal `cfg` affect training at all or is that what all the `setup/change_*` methods are for? I'm assuming the latter ofc, but making sure.

**Comment by titu1994** [comment]
The change* methods are to change the subparts of the model itself - the data loaders, the optimizer/schedule and any extra things - but they don't update the config. The config details how the model *was* built, not how it currently *is* built. If you update the config then the model is rebuilt the next time with the updated info.

**Comment by piraka9011** [comment]
Unfortunately, setting `model.cfg = cfg` before `trainer.fit()` did not change the model cfg in W&B :/

**Comment by titu1994** [comment]
Colab seems to work fine, the wandb report does show different hyper parameters. Locally I could get it working as well with PTL 1.5.5 and main branch

**Comment by piraka9011** [comment]
Using `main` branch still does not work.

```py
asr_model = EncDecRNNTBPEModel.from_pretrained(model_name=cfg.init_from_pretrained_model)
asr_model.cfg = cfg.model
asr_model.set_trainer(trainer)
trainer.fit(asr_model)
```

**Comment by piraka9011** [comment]
It also doesn't change things like:
* `loss` and kwargs
* `log_prediction`

**Comment by titu1994** [comment]
Weird. We'll take a look at it.

**Comment by piraka9011** [comment]
Ok, it looks like the model cfg changed but not the W&B config.
You were correct, apologies for the confusion.

Taking a look as to why the W&B cfg is not changing.

**Comment by piraka9011** [comment]
So before `trainer.fit(asr_model)` is called, `wandb.config` is empty.
After training is done, `wandb.config` is populated, but with the cfg from the pretrained model.

**Comment by piraka9011** [comment]
Looks like the config is updated [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/16a04b29eb811f105a11db4f7f3003489252f9f8/pytorch_lightning/trainer/trainer.py#L1164) which calls the `WandbLogger`'s [`log_hyperparams`](https://github.com/PyTorchLightning/pytorch-lightning/blob/1c28f361d47257feb0bd34cbc8d41642a1f0a2ab/pytorch_lightning/loggers/wandb.py#L358)

**Comment by titu1994** [comment]
Thanks for digging into this !  This is very interesting "bug" - it affects cases where you load a model to fine-tune, modify config then save and retrain. It's not an invalid usecase, we should at least notify PTL team of this. 

On the Nemo side I could hack up a solution for PTL 1.6 (inside of the model.cfg setter) and await PTLs official solution for Nemo 1.7 release. 

Setup metric is kinda niche - you'll normally never modify metrics. ASR is weird where the domain changes quite drastically (eng to mandarin for example) and so ofc the calculation metric needs to change entirely (wer to cer). NLP / NMT has a universal metric like accuracy or BLEU which works without modifications. I'll discuss with the ASR team if they thing a specialized setup_metric() would be worth it. Thanks for the feedback !

**Comment by titu1994** [comment]
Closed via #3552
